File: packages/sdk/src/workspaces.ts
import {QelosSDKOptions} from './types';
import BaseSDK from './base-sdk';

export interface IWorkspaceMember {
  user: string;
  roles: string[];
  created?: string | Date;
}

export interface IInvite {
  _id?: string;
  name?: string;
  email: string;
  phone?: string;
  roles?: string[];
  created?: string | Date;
}

export interface IWorkspace {
  name: string;
  logo?: string;
  isPrivilegedUser?: boolean;

  members?: IWorkspaceMember[];
  invites?: IInvite[];
  labels: string[]

  [key: string]: any;
}

export default class QlWorkspaces extends BaseSDK {
  private relativePath = '/api/workspaces';

  constructor(options: QelosSDKOptions) {
    super(options)
  }

  getWorkspace(workspaceId: string) {
    return this.callJsonApi<IWorkspace>(`${this.relativePath}/${workspaceId}`)
  }

  getMembers(workspaceId: string) {
    return this.callJsonApi<IWorkspaceMember[]>(`${this.relativePath}/${workspaceId}/members`)
  }

  getList() {
    return this.callJsonApi<IWorkspace[]>(this.relativePath);
  }

  remove(workspaceId: string): Promise<any> {
    return this.callApi(`${this.relativePath}/${workspaceId}`, {method: 'delete'});
  }

  update(workspaceId: string, changes: Partial<IWorkspace>): Promise<IWorkspace> {
    return this.callJsonApi<IWorkspace>(
      `${this.relativePath}/${workspaceId}`,
      {
        method: 'put',
        headers: {'content-type': 'application/json'},
        body: JSON.stringify(changes)
      }
    )
  }

  create(workspace: Partial<IWorkspace> & { name: string }): Promise<IWorkspace> {
    return this.callJsonApi<IWorkspace>(this.relativePath, {
      method: 'post',
      headers: {'content-type': 'application/json'},
      body: JSON.stringify(workspace)
    })
  }

  activate(workspaceId: string): Promise<IWorkspace> {
    return this.callJsonApi<IWorkspace>(
      `${this.relativePath}/${workspaceId}/activate`,
      {
        method: 'post',
        headers: {'content-type': 'application/json'},
      }
    )
  }

  async inviteUser(
    workspaceId: string,
    email: string,
    roles?: string[]
  ): Promise<IWorkspace> {
    const ws = await this.getWorkspace(workspaceId);
    const inviteRoles = roles && roles.length > 0 ? roles : ['member'];
    const invites: IInvite[] = [...(ws.invites || [])];
    const normalized = email.trim().toLowerCase();
    const index = invites.findIndex((i) => (i.email || '').trim().toLowerCase() === normalized);
    if (index >= 0) {
      invites[index] = {...invites[index], email: invites[index].email || email, roles: inviteRoles};
    } else {
      invites.push({email, roles: inviteRoles});
    }
    return this.update(workspaceId, {invites});
  }

  removeMember(workspaceId: string, userId: string): Promise<{
    message: string;
    removedMemberId?: string;
    userId?: string;
  }> {
    const path = `${this.relativePath}/${workspaceId}/members/${encodeURIComponent(userId)}`;
    return this.callJsonApi(path, {method: 'delete'});
  }

  updateMemberRoles(
    workspaceId: string,
    userId: string,
    roles: string[]
  ): Promise<{
    message: string;
    updatedMember?: IWorkspaceMember;
    workspaceId?: string;
  }> {
    const path = `${this.relativePath}/${workspaceId}/members/${encodeURIComponent(userId)}`;
    return this.callJsonApi(path, {
      method: 'put',
      headers: {'content-type': 'application/json'},
      body: JSON.stringify({roles}),
    });
  }

  async listInvites(workspaceId: string): Promise<IInvite[]> {
    const ws = await this.getWorkspace(workspaceId);
    return ws.invites ?? [];
  }

  async revokeInvite(workspaceId: string, inviteId: string): Promise<IWorkspace> {
    const ws = await this.getWorkspace(workspaceId);
    const prev = ws.invites || [];
    const idStr = String(inviteId);
    const invites = prev.filter((inv) => String(inv._id ?? '') !== idStr);
    if (invites.length === prev.length) {
      throw new Error('Invitation not found');
    }
    return this.update(workspaceId, {invites});
  }
}

File: packages/python-sdk/qelos_sdk/workspaces.py
from __future__ import annotations

import json
from typing import Any, Dict, List, Optional

from .base_sdk import BaseSDK
from .types import QelosSDKOptions


class QlWorkspaces(BaseSDK):
    """Workspace management: create, list, update, remove, and activate workspaces."""

    _relative_path = "/api/workspaces"

    def __init__(self, options: QelosSDKOptions) -> None:
        super().__init__(options)

    async def get_workspace(self, workspace_id: str) -> Dict[str, Any]:
        """Get a workspace by ID."""
        return await self.call_json_api(f"{self._relative_path}/{workspace_id}")

    async def get_members(self, workspace_id: str) -> List[Dict[str, Any]]:
        """Get members of a workspace."""
        return await self.call_json_api(f"{self._relative_path}/{workspace_id}/members")

    async def get_list(self) -> List[Dict[str, Any]]:
        """List all workspaces for the current user."""
        return await self.call_json_api(self._relative_path)

    async def remove(self, workspace_id: str) -> Any:
        """Delete a workspace."""
        return await self.call_api(f"{self._relative_path}/{workspace_id}", method="DELETE")

    async def update(self, workspace_id: str, changes: Dict[str, Any]) -> Dict[str, Any]:
        """Update a workspace."""
        return await self.call_json_api(
            f"{self._relative_path}/{workspace_id}",
            method="PUT",
            headers={"content-type": "application/json"},
            body=json.dumps(changes),
        )

    async def create(self, workspace: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new workspace. Must include a ``name`` field."""
        return await self.call_json_api(
            self._relative_path,
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(workspace),
        )

    async def activate(self, workspace_id: str) -> Dict[str, Any]:
        """Activate a workspace."""
        return await self.call_json_api(
            f"{self._relative_path}/{workspace_id}/activate",
            method="POST",
            headers={"content-type": "application/json"},
        )
